File: built-in plugins dsp/mb-strings-pizzicato.cpp
/**
 * MB Pizzicato
 * Category : instrument
 * Type     : strings
 * Version  : 1.0.0
 * Author   : Max Booster
 * Desc     : Plucked string pizzicato
 *
 * Part of Max Booster Built-In Plugins DSP
 */

#ifndef MB_STRINGS_PIZZICATO_H
#define MB_STRINGS_PIZZICATO_H

#include <algorithm>
#include <cmath>
#include <cstring>
#include "PluginBase.h"

class MbStringsPizzicato : public PluginBase {
public:
    static constexpr const char* PLUGIN_ID      = "mb-strings-pizzicato";
    static constexpr const char* PLUGIN_NAME    = "MB Pizzicato";
    static constexpr const char* PLUGIN_TYPE    = "strings";
    static constexpr const char* PLUGIN_CATEGORY = "instrument";
    static constexpr const char* VERSION         = "1.0.0";

    struct Parameters {
    float pluck = 0.8f;  // range [0, 1]
    float volume = 0.8f;  // range [0, 1]
    };

    MbStringsPizzicato() = default;
    ~MbStringsPizzicato() override = default;

    void setSampleRate(double sampleRate) override {
        sampleRate_ = sampleRate;
        reset();
    }

    void reset() override {
        std::memset(buffer_, 0, sizeof(buffer_));
    }

    void process(float* left, float* right, int numSamples, Parameters params) {
        params.pluck = std::clamp(params.pluck, 0f, 1f);
        params.volume = std::clamp(params.volume, 0f, 1f);
        for (int i = 0; i < numSamples; ++i) {
            left[i]  = processSample(left[i],  params);
            right[i] = processSample(right[i], params);
        }
    }

private:
    double sampleRate_ = 44100.0;
    float  buffer_[65536] = {};

    inline float processSample(float input, const Parameters& params) {
        // DSP implementation for MB Pizzicato
        return input;
    }
};

#endif // MB_STRINGS_PIZZICATO_H

File: built-in plugins dsp/mb-strings-pizzicato.ts
import type { PluginDefinition } from '../server/services/pluginHostService';

const MbStringsPizzicatoPlugin: PluginDefinition = {
    id: 'mb-strings-pizzicato', slug: 'mb-strings-pizzicato', name: 'MB Pizzicato', category: 'instrument', type: 'strings', version: '1.0.0',
    description: 'Plucked string pizzicato', author: 'Max Booster', grade: 'A',
    oscillators: [{ type: 'triangle', detune: 0, gain: 0.7 }, { type: 'sine', detune: 1200, gain: 0.3 }],
    envelope: { attack: 0.001, decay: 0.3, sustain: 0.0, release: 0.2 },
    parameters: [
      { id: 'pluck', name: 'Pluck', type: 'float', defaultValue: 0.8, minValue: 0, maxValue: 1, automatable: true },
      { id: 'volume', name: 'Volume', type: 'float', defaultValue: 0.8, minValue: 0, maxValue: 1, automatable: true },
    ],
    defaultPreset: { pluck: 0.8, volume: 0.8 },
  };

export default MbStringsPizzicatoPlugin;
